File: leetcode/removeElements.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeElements(
            self, head: Optional[ListNode],
            val: int) -> Optional[ListNode]:
        return remove(head, val)


def remove(head, val):
    if not head:
        return head

    while head and head.val == val:
        head = head.next
        cur = head
        cur.next = None

    cur = head
    while cur:
        next = cur.next
        if next.val == val:
            cur.next = next.next
            next.next = None
        else:
            cur = next
    return head

File: leetcode/removeElements.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 */
 var removeElements = function(head, val) {
  while (head && head.val === val) {
    let cur = head;
    head = head.next;
    cur.next = null;
  }
  if(!head) return head;
  let cur = head;
  while(cur.next) {
    let next = cur.next;
    if(next.val === val) {
      cur.next = next.next;
      next.next = null;
    } else {
      cur = cur.next;
    }
  }
  return head;
};